# Pandas I | Basics

Pandas (from "Panel Data') is an essential piece of scientific (and not only) data analysis infrastructure. It is, in essence, an highly optimized library for manipulating very large tables (or "Data Frames"). 

Today we will be using a single notebook that explains basics of this powerful tool:

[Pandas 1](https://colab.research.google.com/github/nekrut/BMMB554/blob/master/2021/ipynb/Pandas1.ipynb) - Basics using SARS-CoV-2 SRA data. 

## Pandas learning resources

- [Getting started](https://pandas.pydata.org/docs/getting_started/index.html#getting-started) - official introduction from Pandas.
- [Data Science Tools](http://people.duke.edu/~ccc14/bios-821-2017/index.html) - Data Science for Biologists from Duke University.
- [Data Carpentry](https://datacarpentry.org/) - a collection of lessons *à la* Software Carpentry.

## A bit about SARS-CoV-2 data

SRA at NCBI contains the majority of publicly available sequencing datasets produced by various worldwide research efforts. As of late January 2021 sequence read archive (SRA) at the US National Center for Biotechnology Information (NCBI) contained 190,288 raw read datasets summarized here:

> ###### Number of datasets (_N_) and studies (_P_) per platform and strategy. 


| Library Strategy    | Platform        |      N |   P |
|:--------------------|:----------------|-------:|----:|
| AMPLICON            | BGISEQ          |     21 |   1 |
| RNA-Seq             | BGISEQ          |      1 |   1 |
| WGA                 | BGISEQ          |      1 |   1 |
| AMPLICON            | CAPILLARY       |      7 |   1 |
| AMPLICON            | ILLUMINA        | 149668 |  61 |
| WGS                 | ILLUMINA        |   6201 |  43 |
| RNA-Seq             | ILLUMINA        |   4434 |  33 |
| Targeted-Capture    | ILLUMINA        |   1690 |  11 |
| WGA                 | ILLUMINA        |    377 |   4 |
| OTHER               | ILLUMINA        |    148 |  13 |
| AMPLICON            | ION_TORRENT     |    435 |   7 |
| RNA-Seq             | ION_TORRENT     |     42 |   4 |
| WGS                 | ION_TORRENT     |     33 |   6 |
| AMPLICON            | OXFORD_NANOPORE |  25754 |  32 |
| WGS                 | OXFORD_NANOPORE |    936 |  12 |
| WGA                 | OXFORD_NANOPORE |    580 |   3 |
| RNA-Seq             | OXFORD_NANOPORE |     10 |   5 |
| OTHER               | OXFORD_NANOPORE |      4 |   1 |
| AMPLICON            | PACBIO_SMRT     |     12 |   1 |
| Synthetic-Long-Read | PACBIO_SMRT     |      2 |   1 |



While SRA metadata is not perfect (e.g., term `WGA`, whole genome amplification, is likely equivalent to `Amplicon`) the table shows three primary types of data: (1) Illumina-based Ampliconic, (2) Oxford nanopore (ONT)-based Ampliconic, and (3) Illumina-based RNASeq. Of these three types of studies Illumina-based RNASeq is the most suitable for accurate assessment of the intra-host variability.  This approach avoids amplification biases characteristic of PCR-based enrichment approaches such as [PrimalSeq](http://dx.doi.org/10.1371/journal.pone.0239403)—a primary methodology used to generate Illumina- and ONT-based ampliconic datasets.  

